CPP_STYLE_GUIDE.md

```diff
@@ -200,16 +203,29 @@ void FooConsumer(std::unique_ptr<Foo> ptr);
 
 Never use `std::auto_ptr`. Instead, use `std::unique_ptr`.
 
+### No non-const references
+
+Using non-const references often obscures which values are changed by an
+assignment. A pointer is almost always a better choice.
+
 ## Others
 
-## Type casting
+### Type casting
 
 - Always avoid C-style casts (`(type)value`)
 - `dynamic_cast` does not work because RTTI is not enabled
 - Use `static_cast` for casting whenever it works
 - `reinterpret_cast` is okay if `static_cast` is not appropriate
 
-## Do not include `*.h` if `*-inl.h` has already been included
+### Using `auto`
+
+Being explicit about types is preferred over using `auto`.
```

Comment:
Yes please!

Reply:
One of the main benefits of adopting an established style guide is to reduce bikeshedding. I recommend the following talk by Herb Sutter (head of the c++ ISO group) https://www.youtube.com/watch?v=hEx5DNLWGgA

Reply:
I've also found `auto` convenient and helpful when the type is obvious from context... e.g. 

```c++
auto traced_value = tracing::TracedValue::Create();
```

Reply:
@refack Just realized I watched the same argument made by Herb Sutter in 2014 [back in 2015](https://github.com/joyeecheung/my-tech-diary/blob/3aa5de496e3d45db994ae77d936fa4fe3c14ed27/2015/08/2015-08-18.md) but still my experience in real-world code bases taught me otherwise...not that I am completely against `auto`, though, but I still prefer explicit types when it helps with readability.

Reply:
@jasnell Even in that example one would have to look up what `Create()` actually returns… for example, it could definitely be either a smart pointer or a regular pointer.

Reply:
> I think the point is to reduce "clutter" and allow the reader to focus on the intent of the code. 

There are other measures, though – in particular, correctness is an important one. By requesting more explicit types, we put that before focusing on intent (as in this example).

> After all that is what 100% of code in JS looks like and we manage excellently.

If that were completely true, things like TypeScript wouldn’t exist – which is something that’s used by [46 % of npm users](https://twitter.com/seldo/status/1024052940355526656), after all, and the added correctness guarantees are definitely a major reason for that.

Reply:
@refack 

> After all that is what 100% of code in JS looks like and we manage excellently.

Not necessarily, for properties and a lot of local variables we initialize them upfront so in the declaration it's already very clear what their types are supposed to be. Personally I would want to see more JSDoc-like annotations in `lib` (not necessary for simple helpers, but would be very helpful for functions with a lot of arguments named with something like `obj`, `value`, `handle`, .etc).